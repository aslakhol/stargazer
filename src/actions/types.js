// UI State
export const OPEN_MODAL = 'OPEN_MODAL';
export const CLOSE_MODAL = 'CLOSE_MODAL';

// Data Related
export const SET_CURRENT_PERSON = 'SET_CURRENT_PERSON';

// Query related
export const REQUEST_PERSON = 'REQUEST_PERSON';
export const RECIEVE_PERSON = 'RECIEVE_PERSON';
export const NEW_QUERY = 'NEW_QUERY';
export const NEW_TIMEOUT = 'NEW_TIMEOUT';
