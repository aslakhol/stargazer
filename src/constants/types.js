export const NEW_QUERY = "NEW_QUERY";
export const REQUEST_PERSON = "ADD_PERSON";
export const RECIEVE_PERSON = "RECIEVE_PERSON";
