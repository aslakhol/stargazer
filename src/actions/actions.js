import fetch from 'cross-fetch';

export const REQUEST_PERSON = 'REQUEST_PERSON';
export const RECIEVE_PERSON = 'RECIEVE_PERSON';
export const NEW_QUERY = 'NEW_QUERY';
export const SELECT_PERSON = 'SELECT_PERSON';


export const selectPerson = person => ({
  type: SELECT_PERSON,
  person,
});

const requestPerson = person => ({
  type: REQUEST_PERSON,
  person,
});

const recievePerson = (person, json) => ({
  type: RECIEVE_PERSON,
  person: json.data.children.map(child => child.data),
  recievedAt: Date.now(),
});

const fetchPersons = person => (dispatch) => {
  dispatch(requestPerson(person));
  return fetch(`https://swapi.co/api/people?search=${person}.json`)
    .then(response => response.json())
    .then(json => dispatch(recievePerson(person, json)));
};

const shouldFetch = (state, person) => {
  const persons = state.personsByPerson[person];
  if (!persons) {
    return true;
  } if (persons.isFetching) {
    return false;
  }
  return persons;
};

export const fetchPersonsIfNeeded = person => (dispatch, getState) => {
  if (shouldFetch(getState(), person)) {
    return dispatch(fetchPersons(person));
  }
};
