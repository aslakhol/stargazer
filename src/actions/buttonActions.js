import { INCLUDE_SELECTED, EXCLUDE_SELECTED, SORT_BY_SELECTED } from '../utils/constants';

export const includeSelectedInQuery = selected => ({ type: INCLUDE_SELECTED, selected });
export const excludeSelectedInQuery = selected => ({ type: EXCLUDE_SELECTED, selected });
export const sortBySelected = selected => ({ type: SORT_BY_SELECTED, selected });
