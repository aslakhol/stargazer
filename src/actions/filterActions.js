import { INCLUDE_SELECTED, EXCLUDE_SELECTED } from '../utils/constants';

export const includeSelectedInQuery = selected => ({ type: INCLUDE_SELECTED, selected });
export const excludeSelectedInQuery = selected => ({ type: EXCLUDE_SELECTED, selected });
