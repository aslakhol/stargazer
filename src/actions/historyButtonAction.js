import { SHOW_HISTORY, HIDE_HISTORY } from '../constants';

export const showHistory = () => ({ type: SHOW_HISTORY });
export const hideHistory = () => ({ type: HIDE_HISTORY });
