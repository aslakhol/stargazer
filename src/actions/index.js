import { NEW_QUERY, REQUEST_PERSON, RECIEVE_PERSON } from "../constants/types"

const initialState = {
  items: [],
}

const updateQuery = (newQuery) => {
  return {
    type: NEW_QUERY,
    newQuery
  };
}

const requestPerson = (newQuery) => {
  return {
    type: REQUEST_PERSON,
    newQuery,
  };
}

const recievePerson = (newQuery, persons) => {
  return {
    type: RECIEVE_PERSON,
    newQuery,
    persons,
  };
}

const items  = (array) =>{
  let personArray = [];
  array.forEach((item) => {
    personArray = personArray.concat(item.results);
  });
  return personArray.map((item) => {
    return {
      type: "person",
      name: item.name,
      gender: item.gender,
      height: item.height,
      mass: item.mass,
    };
  })
}

const fetchPersons = (newQuery) => {
  const endpoint = [
    `https://swapi.co/api/people?search=${newQuery}`,
  ];

  return (dispatch) => {
    dispatch(updateQuery(newQuery));
    dispatch(requestPerson(newQuery));
    return (endpoint.map(url => 
      fetch(url).then(resp => resp.json()),
    )).then(json => console.log(json))
    .then(array => items(array))
    .then(json => dispatch(recievePerson(newQuery, json)));
  };
}

const shouldFetch  = (state, newQuery) => {
  const stuff = state.itemsByQuery[newQuery];
  if (!stuff){
    return true;
  } else if (stuff.isFetching){
    return false;
  }
  return false;
}

export const fetchItems = (newQuery) => {
  return(dispatch, getState) => {
    if(shouldFetch(getState(), newQuery)){
      return dispatch(fetchPersons(newQuery));
    }
    return dispatch(updateQuery(newQuery));
  }
}
