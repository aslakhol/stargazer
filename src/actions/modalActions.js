import { OPEN_MODAL, CLOSE_MODAL } from './types';

export const setModalOpen = () => ({ type: OPEN_MODAL });
export const setModalClosed = () => ({ type: CLOSE_MODAL });
