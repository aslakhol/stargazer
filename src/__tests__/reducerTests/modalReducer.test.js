import modalReducer from '../../reducers/modalReducer';
import { OPEN_MODAL, CLOSE_MODAL } from '../../actions/types';

describe('modalReducer', () => {
  describe('INITIAL_STATE', () => {
    it('is correct', () => {
      const action = { type: 'DUMMY_ACTION' };
      const initialState = false;

      expect(modalReducer(undefined, action)).toEqual(initialState);
    });
  });

  describe('OPEN_MODAL', () => {
    it('returns the correct state', () => {
      const action = { type: OPEN_MODAL, data: true };
      expect(modalReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('CLOSE_MODAL', () => {
    it('returns the correct state', () => {
      const action = { type: CLOSE_MODAL, data: false };
      expect(modalReducer(undefined, action)).toMatchSnapshot();
    });
  });
});
