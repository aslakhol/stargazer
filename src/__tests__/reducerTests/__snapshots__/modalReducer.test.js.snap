// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modalReducer CLOSE_MODAL returns the correct state 1`] = `false`;

exports[`modalReducer OPEN_MODAL returns the correct state 1`] = `true`;
